Plugins/DD4hep/src/DD4hepBinningHelpers.cpp

```diff
@@ -0,0 +1,131 @@
+      // Set the number of bins
+      variantParams.set<int>(bname + "_" + bv + "_n", nBins);
+      // Set min/max paraeter
+      if (!autoRange) {
+        variantParams.set<double>(
+            bname + "_" + bv + "_min",
+            xmlBinning.attr<double>(std::string(bv + "min").c_str()));
+        variantParams.set<double>(
+            bname + "_" + bv + "_max",
+            xmlBinning.attr<double>(std::string(bv + "max").c_str()));
+      }
+    } else {
+      // Variable binning detected
+      variantParams.set<std::string>(bname + "_" + bv + "_type", "variable");
+      // Get the number of bins explicitly
+      auto boundaries =
+          xmlBinning.attr<std::string>(std::string(bv + "boundaries").c_str());
+      std::string del = ",";
+      int end = boundaries.find(del);
+      int ib = 0;
+      // Unit conversion
+      double unitScalar = 1.;
+      if (bv != "phi") {
+        unitScalar = Acts::UnitConstants::mm / dd4hep::millimeter;
+      }
+      // Split and convert
+      while (end != -1) {
+        double bR = unitScalar * dd4hep::_toFloat(boundaries.substr(0, end));
+        variantParams.set<double>(
+            bname + "_" + bv + "_b" + std::to_string(ib++), bR);
+        boundaries.erase(boundaries.begin(), boundaries.begin() + end + 1);
+        end = boundaries.find(del);
+      }
+      double bR = unitScalar * std::stod(boundaries.substr(0, end));
+      variantParams.set<double>(bname + "_" + bv + "_b" + std::to_string(ib),
+                                bR);
+      // The number of bins are needed to unpack the data
+      variantParams.set<int>(bname + "_" + bv + "_n", ib);
+    }
+  }
+}
+
+std::vector<Acts::Experimental::ProtoBinning>
+Acts::DD4hepBinningHelpers::convertBinning(
+    const dd4hep::DetElement &dd4hepElement, const std::string &bname) {
+  // Return proto binning vector
+  std::vector<Experimental::ProtoBinning> protoBinnings;
+
+  for (const auto &[ab, bVal] : allowedBinnings) {
+    auto type =
+        getParamOr<std::string>(bname + "_" + ab + "_type", dd4hepElement, "");
+    if (!type.empty()) {
+      // Default binning is bound
+      auto bType = Acts::detail::AxisBoundaryType::Bound;
+      // Equidistant or variable binning
+      detail::AxisType aType = type == "equidistant"
+                                   ? detail::AxisType::Equidistant
+                                   : detail::AxisType::Variable;
+      int nBins = getParamOr<int>(bname + "_" + ab + "_n", dd4hepElement, 0);
+      int nExpansion =
+          getParamOr<int>(bname + "_" + ab + "_exp", dd4hepElement, 0);
+      // Indicate auto-range checking
+      bool autoRange = getParamOr<bool>(bname + "_" + ab + "_autorange",
+                                        dd4hepElement, false);
+      // Equidistant binning
+      if (aType == detail::AxisType::Equidistant) {
+        if (autoRange) {
+          protoBinnings.push_back(
+              Experimental::ProtoBinning(bVal, bType, nBins, nExpansion));
```

Comment:
Ok so for my understanding, proto binning indicates auto binning based on discovered surfaces?

Reply:
The `ProtoBinning` class is basically the stripped out (without actual bin calculation) binning description, which we used to do with the `BinUtility`. The latter is not used in the new `Detector` schema, as the actual bin lookup and storage is all done via grids.

One option of the `ProtoBinning` is to define the binning without knowing the edges yet (this practically only makes sense for equidistant), e.g. you want to say: I want 100 material bins on this portal surface in `z`, when you do not know the actual necessary z boundary yet.